File: python/src/shadow/sdk/session.py
"""Session — the recording context manager that writes an `.agentlog` file.

Usage (auto-instrumentation, default):

```python
from shadow.sdk import Session
import anthropic

with Session(output_path="trace.agentlog", tags={"env": "dev"}):
    client = anthropic.Anthropic()
    response = client.messages.create(
        model="claude-opus-4-7",
        max_tokens=1024,
        messages=[{"role": "user", "content": "hello"}],
    )
    # Shadow records this call automatically; no record_chat needed.
```

The auto-instrumentor monkey-patches `anthropic.resources.messages.{,Async}Messages.create`
and `openai.resources.chat.completions.{,Async}Completions.create` at
`Session.__enter__` and restores them at `__exit__`. If an SDK isn't
installed, it's skipped cleanly. Streaming calls pass through un-recorded.

You can also record manually — useful for custom backends or reshaping:

```python
with Session(output_path="trace.agentlog") as session:
    session.record_chat(request_dict, response_dict)
```

Pass `auto_instrument=False` to disable the monkey-patching.
"""

from __future__ import annotations

import contextlib
import datetime
import os
import platform
import sys
from pathlib import Path
from types import TracebackType
from typing import Any, Self

from shadow import __version__, _core
from shadow.redact import Redactor


def _now_iso() -> str:
    """RFC 3339 UTC timestamp with millisecond precision, ending in `Z`."""
    now = datetime.datetime.now(datetime.UTC)
    return now.strftime("%Y-%m-%dT%H:%M:%S.") + f"{now.microsecond // 1000:03d}Z"


class Session:
    """Record LLM interactions into a `.agentlog` file.

    Parameters
    ----------
    output_path:
        Path to write the `.agentlog` file on exit. Parent directories
        are created if missing.
    tags:
        Key→value string tags placed in the metadata record's
        `payload.tags`. Indexed by the SQLite store in CLI workflows.
    session_tag:
        Free-form label placed in `envelope.meta.session_tag`. Used for
        the SQLite index's `session_tag` column.
    redactor:
        Optional custom redactor. Defaults to [`Redactor()`] (SPEC §9
        defaults). Pass `None` to disable (NOT recommended).
    """

    def __init__(
        self,
        output_path: Path | str,
        tags: dict[str, str] | None = None,
        session_tag: str | None = None,
        redactor: Redactor | None = None,
        auto_instrument: bool = True,
        trace_id: str | None = None,
        parent_span_id: str | None = None,
    ) -> None:
        from shadow.sdk.tracing import current_parent_span_id, current_trace_id, new_trace_id

        self._output_path = Path(output_path)
        self._tags = dict(tags or {})
        self._session_tag = session_tag
        self._redactor: Redactor | None = redactor if redactor is not None else Redactor()
        self._records: list[dict[str, Any]] = []
        self._root_id: str | None = None
        self._auto_instrument = auto_instrument
        self._instrumentor: Any | None = None
        # Distributed-trace propagation: inherit trace_id from the env if a
        # parent set it, else mint a new one. span_id is this session's own.
        self._trace_id = trace_id or current_trace_id() or new_trace_id()
        self._parent_span_id = parent_span_id or current_parent_span_id()

    def __enter__(self) -> Self:
        meta_payload: dict[str, Any] = {
            "sdk": {"name": "shadow", "version": __version__},
            "runtime": {
                "python": sys.version.split()[0],
                "platform": f"{platform.system().lower()}-{platform.machine().lower()}",
            },
        }
        if self._tags:
            meta_payload["tags"] = dict(self._tags)
        meta_payload = self._redact(meta_payload)
        meta_id = _core.content_id(meta_payload)
        self._root_id = meta_id
        self._records.append(self._envelope("metadata", meta_payload, meta_id, parent=None))
        if self._auto_instrument:
            from shadow.sdk.instrumentation import Instrumentor

            # If install() fails mid-way, uninstall whatever patches ARE
            # recorded so we don't leave the global SDK classes in a
            # half-patched state across the session boundary.
            instr = Instrumentor(self)
            try:
                instr.install()
            except Exception:
                with contextlib.suppress(Exception):
                    instr.uninstall()
                raise
            self._instrumentor = instr
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None:
        if self._instrumentor is not None:
            self._instrumentor.uninstall()
            self._instrumentor = None
        self._output_path.parent.mkdir(parents=True, exist_ok=True)
        bytes_out = _core.write_agentlog(self._records)
        self._output_path.write_bytes(bytes_out)

    @property
    def root_id(self) -> str | None:
        """Root record id, available after __enter__."""
        return self._root_id

    @property
    def trace_id(self) -> str:
        """Distributed trace id. Same across all sessions linked via env vars."""
        return self._trace_id

    def env_for_child(self) -> dict[str, str]:
        """Env dict to pass to a child process so it joins this trace."""
        from shadow.sdk.tracing import env_for_child

        # Use the root record's id as this session's span_id for clarity.
        span_id = (self._root_id or "").replace("sha256:", "")[:16]
        return env_for_child(self._trace_id, span_id or "0" * 16)

    def record_chat(
        self,
        request: dict[str, Any],
        response: dict[str, Any],
        parent_id: str | None = None,
    ) -> tuple[str, str]:
        """Append a chat_request + chat_response pair to the trace.

        Returns `(request_id, response_id)`. The caller can use them for
        subsequent `record_tool_call` / `record_tool_result` parent refs.
        """
        if self._root_id is None:
            raise RuntimeError("Session not entered; use `with Session(...) as s:`")
        parent = parent_id if parent_id is not None else self._last_id()
        redacted_req = self._redact(request)
        req_id = _core.content_id(redacted_req)
        req_rec = self._envelope("chat_request", redacted_req, req_id, parent=parent)
        self._records.append(req_rec)
        redacted_resp = self._redact(response)
        resp_id = _core.content_id(redacted_resp)
        resp_rec = self._envelope("chat_response", redacted_resp, resp_id, parent=req_id)
        self._records.append(resp_rec)
        return req_id, resp_id

    def record_tool_call(
        self,
        tool_name: str,
        tool_call_id: str,
        arguments: dict[str, Any],
        parent_id: str | None = None,
    ) -> str:
        payload = {
            "tool_name": tool_name,
            "tool_call_id": tool_call_id,
            "arguments": arguments,
        }
        payload = self._redact(payload)
        record_id = _core.content_id(payload)
        parent = parent_id if parent_id is not None else self._last_id()
        self._records.append(self._envelope("tool_call", payload, record_id, parent=parent))
        return record_id

    def record_tool_result(
        self,
        tool_call_id: str,
        output: str | dict[str, Any],
        is_error: bool = False,
        latency_ms: int = 0,
        parent_id: str | None = None,
    ) -> str:
        payload = {
            "tool_call_id": tool_call_id,
            "output": output,
            "is_error": is_error,
            "latency_ms": latency_ms,
        }
        payload = self._redact(payload)
        record_id = _core.content_id(payload)
        parent = parent_id if parent_id is not None else self._last_id()
        self._records.append(self._envelope("tool_result", payload, record_id, parent=parent))
        return record_id

    def _envelope(
        self, kind: str, payload: dict[str, Any], record_id: str, parent: str | None
    ) -> dict[str, Any]:
        env: dict[str, Any] = {
            "version": _core.SPEC_VERSION,
            "id": record_id,
            "kind": kind,
            "ts": _now_iso(),
            "parent": parent,
            "payload": payload,
        }
        meta: dict[str, Any] = {}
        if self._session_tag is not None:
            meta["session_tag"] = self._session_tag
        if self._redactor and self._redactor.last_modified:
            meta["redacted"] = True
        meta["trace_id"] = self._trace_id
        if self._parent_span_id is not None:
            meta["parent_span_id"] = self._parent_span_id
        env["meta"] = meta
        return env

    def _redact(self, value: dict[str, Any]) -> dict[str, Any]:
        if self._redactor is None:
            return value
        redacted: dict[str, Any] = self._redactor.redact_value(value)
        return redacted

    def _last_id(self) -> str | None:
        return self._records[-1]["id"] if self._records else None


# Convenience: the SHADOW_SESSION_OUTPUT environment variable is honored by
# `shadow record -- <cmd>` to point the wrapped process at a trace path.
def output_path_from_env() -> Path | None:
    """Return `$SHADOW_SESSION_OUTPUT` if set, else None."""
    env = os.environ.get("SHADOW_SESSION_OUTPUT")
    return Path(env) if env else None

File: typescript/src/session.ts
/** Shadow Session — records chat pairs into a `.agentlog` file. */

import { appendFileSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

import { canonicalJson, contentId, type AgentlogRecord } from './agentlog.js';
import {
  autoInstrument,
  uninstrument,
  type InstrumentorHandle,
} from './instrumentation.js';
import { Redactor } from './redact.js';
import {
  currentParentSpanId,
  currentTraceId,
  envForChild as envForChildImpl,
  newTraceId,
} from './tracing.js';

export interface SessionOptions {
  outputPath: string;
  tags?: Record<string, string>;
  sessionTag?: string;
  autoInstrument?: boolean;
  traceId?: string;
  parentSpanId?: string;
  /** Default Redactor is installed if omitted. Pass `null` to DISABLE redaction
   *  (not recommended — every record's payload is hashed as-is, so any PII or
   *  API keys land in the content-addressed store and persist). */
  redactor?: Redactor | null;
}

const SPEC_VERSION = '0.1';

// Package version — read from the shipped package.json so SDK-provenance
// records carry the actual installed version, not a stale literal. Reading
// at module-load time (cheap, once per process) is standard practice for
// Node SDKs that emit version headers.
function _readSdkVersion(): string {
  try {
    // Walk up from the compiled file location (dist/session.js) to the
    // package root, which contains package.json for this package.
    const here = dirname(fileURLToPath(import.meta.url));
    for (const candidate of [
      `${here}/../package.json`,
      `${here}/../../package.json`,
    ]) {
      try {
        const raw = readFileSync(candidate, 'utf-8');
        const parsed = JSON.parse(raw) as { version?: string; name?: string };
        if (parsed.name === 'shadow-diff' && typeof parsed.version === 'string') {
          return parsed.version;
        }
      } catch {
        continue;
      }
    }
  } catch {
    // fall through
  }
  return '0.0.0';
}
const SDK_VERSION: string = _readSdkVersion();

export class Session {
  private readonly outputPath: string;
  private readonly tags: Record<string, string>;
  private readonly sessionTag: string | undefined;
  private readonly autoInstrumentFlag: boolean;
  private readonly traceIdValue: string;
  private readonly parentSpanId: string | null;
  private readonly redactor: Redactor | null;

  private records: AgentlogRecord[] = [];
  private rootId: string | null = null;
  private instrumentor: InstrumentorHandle | null = null;

  constructor(opts: SessionOptions) {
    this.outputPath = opts.outputPath;
    this.tags = { ...(opts.tags ?? {}) };
    this.sessionTag = opts.sessionTag;
    this.autoInstrumentFlag = opts.autoInstrument ?? true;
    this.traceIdValue = opts.traceId ?? currentTraceId() ?? newTraceId();
    this.parentSpanId = opts.parentSpanId ?? currentParentSpanId();
    // `undefined` → default Redactor; explicit `null` → no redaction.
    this.redactor =
      opts.redactor === undefined ? new Redactor() : opts.redactor;
  }

  async enter(): Promise<this> {
    const metaPayload: Record<string, unknown> = {
      sdk: { name: 'shadow', version: SDK_VERSION },
      runtime: {
        node: process.version,
        platform: `${process.platform}-${process.arch}`,
      },
    };
    if (Object.keys(this.tags).length > 0) {
      metaPayload.tags = { ...this.tags };
    }
    const redactedMeta = this.redact(metaPayload);
    const id = contentId(redactedMeta);
    this.rootId = id;
    // Initialise the on-disk file synchronously and write the metadata
    // record immediately. Doing this in `enter()` (rather than buffering
    // until `exit()`) means a process that crashes mid-run still leaves
    // a parseable .agentlog containing every record produced before the
    // crash, instead of writing nothing at all. JSONL appends are
    // self-terminating so a partial trace parses cleanly.
    mkdirSync(dirname(this.outputPath), { recursive: true });
    const env = this.envelope('metadata', redactedMeta, id, null);
    this.records.push(env);
    writeFileSync(this.outputPath, this._encodeRecord(env));
    if (this.autoInstrumentFlag) {
      this.instrumentor = await autoInstrument(this);
    }
    return this;
  }

  async exit(): Promise<void> {
    if (this.instrumentor) {
      uninstrument(this.instrumentor);
      this.instrumentor = null;
    }
    // No final flush needed — every record was appended synchronously
    // as it was captured (see `enter()` and `recordChat()`). exit() is
    // now solely about un-monkey-patching the openai/anthropic SDKs.
  }

  /** Append a chat_request + chat_response pair. */
  recordChat(
    request: Record<string, unknown>,
    response: Record<string, unknown>,
  ): { requestId: string; responseId: string } {
    if (this.rootId === null) {
      throw new Error('Session not entered; await session.enter() first.');
    }
    const parent = this.lastId();
    // Redact BEFORE content-id so the hash is over the redacted payload,
    // not the original (otherwise PII leaks into the content-address).
    const redactedReq = this.redact(request);
    const redactedResp = this.redact(response);
    const reqId = contentId(redactedReq);
    const reqEnv = this.envelope('chat_request', redactedReq, reqId, parent);
    this.records.push(reqEnv);
    const respId = contentId(redactedResp);
    const respEnv = this.envelope('chat_response', redactedResp, respId, reqId);
    this.records.push(respEnv);
    // Append both records synchronously so a crash AFTER this call
    // returns still leaves them on disk. appendFileSync is one syscall
    // per record and is durable in the same way `>>` from a shell
    // would be — fine for the < 100 records/sec rate a typical agent
    // produces. If recording were on the hot path of every token we'd
    // batch, but chat_request + chat_response is one pair per LLM call.
    appendFileSync(this.outputPath, this._encodeRecord(reqEnv));
    appendFileSync(this.outputPath, this._encodeRecord(respEnv));
    return { requestId: reqId, responseId: respId };
  }

  /**
   * Encode a single record as canonical JSONL bytes (one line + LF).
   * Same wire format as `writeAgentlog([record])` but allocates one
   * record at a time so per-record appends don't quadratically rebuild
   * the whole file.
   */
  private _encodeRecord(env: AgentlogRecord): Uint8Array {
    const line = canonicalJson(env);
    const out = new Uint8Array(line.length + 1);
    out.set(line, 0);
    out[line.length] = 0x0a; // '\n'
    return out;
  }

  get traceId(): string {
    return this.traceIdValue;
  }

  /** Env object a child process should inherit to join this trace. */
  envForChild(): Record<string, string> {
    const spanId = (this.rootId ?? '').replace('sha256:', '').slice(0, 16);
    return envForChildImpl(this.traceIdValue, spanId || '0'.repeat(16));
  }

  private envelope(
    kind: AgentlogRecord['kind'],
    payload: unknown,
    id: string,
    parent: string | null,
  ): AgentlogRecord {
    const meta: AgentlogRecord['meta'] = {};
    if (this.sessionTag !== undefined) meta.session_tag = this.sessionTag;
    meta.trace_id = this.traceIdValue;
    if (this.parentSpanId !== null) meta.parent_span_id = this.parentSpanId;
    return {
      version: SPEC_VERSION,
      id,
      kind,
      ts: nowIso(),
      parent,
      payload,
      meta,
    };
  }

  private lastId(): string | null {
    return this.records.length > 0
      ? this.records[this.records.length - 1].id
      : null;
  }

  /**
   * Counts of captured records by kind. Used by the auto entrypoint
   * to emit a loud zero-capture warning at session exit.
   */
  recordStats(): { total: number; chatRequests: number; chatResponses: number } {
    let chatRequests = 0;
    let chatResponses = 0;
    for (const r of this.records) {
      if (r.kind === 'chat_request') chatRequests++;
      else if (r.kind === 'chat_response') chatResponses++;
    }
    return { total: this.records.length, chatRequests, chatResponses };
  }

  private redact<T>(value: T): T {
    if (this.redactor === null) return value;
    return this.redactor.redactValue(value);
  }
}

function nowIso(): string {
  const d = new Date();
  const pad = (n: number, w: number = 2): string => n.toString().padStart(w, '0');
  return (
    `${d.getUTCFullYear()}-${pad(d.getUTCMonth() + 1)}-${pad(d.getUTCDate())}T` +
    `${pad(d.getUTCHours())}:${pad(d.getUTCMinutes())}:${pad(d.getUTCSeconds())}.` +
    `${pad(d.getUTCMilliseconds(), 3)}Z`
  );
}
